torch/testing/_internal/common_methods_invocations.py

```diff
@@ -88,28 +118,68 @@ class UnaryUfuncInfo(OpInfo):
+                 handles_extremals=True,  # whether the op correctly handles extremal values (like inf)
+                 handles_complex_extremals=True,  # whether the op correct handles complex extremals (like inf -infj)
                  **kwargs):
         super(UnaryUfuncInfo, self).__init__(name,
                                              dtypes=dtypes,
                                              dtypesIfCPU=dtypesIfCPU,
                                              dtypesIfCUDA=dtypesIfCUDA,
                                              dtypesIfROCM=dtypesIfROCM,
                                              **kwargs)
+        self.ref = ref
+        self.domain = domain
+        self.handles_large_floats = handles_large_floats
+        self.handles_extremals = handles_extremals
+        self.handles_complex_extremals = handles_complex_extremals
 
 L = 20
 M = 10
 S = 5
 
+# Situational skips for tests that fail on particular builds
+_windows_skip = ('.*numerics.*complex.*',) if IS_WINDOWS else tuple()
+_mac_skip = ('.*numerics.*cpu.*complex.*',) if IS_MACOS else tuple()
+_rocm_skip = ('.*numerics.*float32.*',) if TEST_WITH_ROCM else tuple()
+
 # Operator database
 op_db = [
+    # NOTE: CPU complex acos produces incorrect outputs (https://github.com/pytorch/pytorch/issues/42952)
+    UnaryUfuncInfo('acos',
+                   ref=np.arccos,
+                   domain=(-1, 1),
+                   handles_complex_extremals=False,
+                   decorators=(precisionOverride({torch.bfloat16: 1e-1,
+                                                  torch.complex64: 1e-2}),),
+                   tests_to_skip=('.*numerics.*cpu.*complex.*',) + _windows_skip),
     UnaryUfuncInfo('cos',
-                   dtypesIfCUDA=floating_and_complex_types_and(torch.half, torch.bfloat16)),
+                   ref=np.cos,
+                   dtypesIfCUDA=floating_and_complex_types_and(torch.half, torch.bfloat16),
+                   handles_large_floats=False,
+                   decorators=(precisionOverride({torch.bfloat16: 1e-2}),),
+                   tests_to_skip=_windows_skip + _mac_skip + _rocm_skip),
     UnaryUfuncInfo('cosh',
-                   dtypesIfCPU=floating_and_complex_types()),
+                   ref=np.cosh,
+                   dtypesIfCPU=floating_and_complex_types(),
+                   tests_to_skip=_windows_skip + _mac_skip),
+    UnaryUfuncInfo('sin',
+                   ref=np.sin,
+                   handles_large_floats=False,
+                   handles_complex_extremals=False,
+                   decorators=(precisionOverride({torch.bfloat16: 1e-2}),),
+                   tests_to_skip=_windows_skip + _rocm_skip),
+    UnaryUfuncInfo('neg',
+                   ref=np.negative,
+                   dtypes=all_types_and_complex_and(torch.half, torch.bfloat16),
+                   dtypesIfCPU=all_types_and_complex_and(torch.half, torch.bfloat16),
+                   dtypesIfCUDA=all_types_and_complex_and(torch.half)),
```

Comment:
If I'm adding a new function, how am I supposed to work out what the `dtypesIfCPU` setting is? I can think of three possibilities:

1. Cargo cult one of the existing definitions
2. Start with no restrictions and then add restrictions as the test suite empirically fails
3. Read the source code and replicate the dispatcher macro there

I'd prefer us not to be in situation (1), because that makes it really easy for someone to pick the wrong operator definition to cargo cult off of and then whoops, your exclusions don't make any sense for the actual operator in question. It also makes it hard to understand how these declarations get updated as the code changes (which makes (3) a lot more attractive, in this situation).

Reply:
Since the functions to acquire types match our ATen dispatch macros I would expect (3), but you could do (2) if you wanted. The test suite WILL yell at you if you get the types wrong for a native device type.

Reply:
> The test suite WILL yell at you if you get the types wrong for a native device type

Only if you include more than you should, not less than, right?

Reply:
Either way. If you include too few then `test_unsupported_dtypes` will test the dtype that's supported but not listed in the op's info and fail becaue that dtype didn't throw a runtime error when given to the op.

Reply:
FWIW, this functionality is SOOO useful that it would be a huge shame if it only got applied on unary ops ;)